local-cluster/tests/local_cluster.rs

```diff
@@ -4480,6 +4480,220 @@ fn test_slot_hash_expiry() {
+            a_tower.node_pubkey = b_pubkey;
+            // Update common_ancestor_slot because A is still running
+            if let Some(s) = a_tower.last_voted_slot() {
+                common_ancestor_slot = s;
+                info!("New common_ancestor_slot {}", common_ancestor_slot);
+            } else {
+                panic!("A's tower has no votes");
+            }
+            info!("Save as B's tower");
+            save_tower(&b_ledger_path, &a_tower, &b_info.info.keypair);
+            info!("B's new tower: {:?}", a_tower.tower_slots());
+        } else {
+            panic!("A's tower is missing");
+        }
+
+        // Get rid of any slots past common_ancestor_slot
+        info!("Removing extra slots from B's blockstore");
+        let blockstore = open_blockstore(&b_ledger_path);
+        purge_slots_with_count(&blockstore, common_ancestor_slot + 1, 100);
+    }
+
+    info!("Run A on majority fork until it creates its own fork");
+    let mut last_vote_on_a;
+    // Keep A running for a while longer so A has its own fork.
+    loop {
+        last_vote_on_a = wait_for_last_vote_in_tower_to_land_in_ledger(&a_ledger_path, &a_pubkey);
+        if last_vote_on_a > common_ancestor_slot {
+            let blockstore = open_blockstore(&a_ledger_path);
+            info!(
+                "A majority fork: {:?}",
+                AncestorIterator::new(last_vote_on_a, &blockstore).collect::<Vec<Slot>>()
+            );
+            break;
+        }
+        sleep(Duration::from_millis(100));
+    }
+
+    // Kill A and restart B with voting. B should now fork off
+    info!("Killing A");
+    let a_info = cluster.exit_node(&a_pubkey);
+
+    info!("Restarting B");
+    b_info.config.voting_disabled = false;
+    cluster.restart_node(&b_pubkey, b_info, SocketAddrSpace::Unspecified);
+
+    // Allow B to fork now.
+    info!("Allowing B to fork");
+    loop {
+        sleep(Duration::from_millis(200));
+        let blockstore = open_blockstore(&b_ledger_path);
+        let (last_vote, _) = last_vote_in_tower(&b_ledger_path, &b_pubkey).unwrap();
+        let mut iterator = AncestorIterator::new(last_vote, &blockstore);
+        let first_ancestor = iterator.next().unwrap();
+        let second_ancestor = iterator.next().unwrap();
+        if second_ancestor > common_ancestor_slot
+            && first_ancestor % 4 == 3
+            && first_ancestor == second_ancestor + 1
+        {
+            break;
+        }
+    }
+
+    info!("Kill B");
+    b_info = cluster.exit_node(&b_pubkey);
+
+    info!("Resolve the partition");
+    {
+        // Here we let B know about the missing blocks that A had produced on its partition
+        let a_blockstore = open_blockstore(&a_ledger_path);
+        let b_blockstore = open_blockstore(&b_ledger_path);
+        copy_blocks(last_vote_on_a, &a_blockstore, &b_blockstore);
+    }
+
+    // Now restart A and B and see if B is able to eventually switch onto the majority fork
```

Comment:
We're not waiting for `B` to switch, just to refresh right?

Reply:
Changed.